src/util/light_map.h

```diff
@@ -0,0 +1,126 @@
+/*******************************************************************\
+
+Module: Memory-efficient unordered_map
+
+Author: Diffblue ltd.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_LIGHT_MAP_H
+#define CPROVER_UTIL_LIGHT_MAP_H
+
+#include <memory>
+#include <vector>
+#include <unordered_map>
+#include "any_ref.h"
+
+/// Copied data is shared with previous instance to avoid copying of elements
+/// For example having a map with 1000 elements, passing it into the
+/// function and modifying one element
+template<typename key_type, typename value_type>
+class light_unordered_mapt
+{
+public:
+  /// Default constructor - constructs empty map
+  light_unordered_mapt():
+    data_(1, std::make_shared<std::unordered_map<key_type, value_type>>())
+  { }
+
+  /// Creates a map that inherits elements with the "other" map
+  light_unordered_mapt(const light_unordered_mapt &other):
+    data_(other.data_)
+  {
+    data_.emplace_back(
+      std::make_shared<std::unordered_map<key_type, value_type>>());
+  }
+
+  /// Move constructor - claim "other"'s map state. Empty "other" map
+  light_unordered_mapt(light_unordered_mapt &&other):
+    light_unordered_mapt()
+  { *this=std::move(other); }
+
+  /// Convert regular map into light map
+  explicit light_unordered_mapt(
+    std::unordered_map<key_type, value_type> &&other):
+    data_(
+      1,
+      std::make_shared<std::unordered_map<key_type, value_type>>(
+        std::move(other)))
+    { }
+
+  /// Copy assignment - Clear this map and inherit from "other" map
+  light_unordered_mapt &operator=(const light_unordered_mapt &other)
+  {
+    light_unordered_mapt copy(other);
+    *this=std::move(copy);
+    return *this;
+  }
+
+  /// Move assignment - swap state with "other" map
+  light_unordered_mapt &operator=(light_unordered_mapt &&other)
+  {
+    std::swap(other.data_, data_);
+    return *this;
+  }
+
+  /// Set value for a specified key
+  void set(const key_type &key, const value_type &value)
+  {
+    auto &map=*data_.back();
+    auto it=map.find(key);
```

Comment:
Can't you write both `set` functions as:
```c++
*data.back()[key_type] = value; // other version needs std::move here
```

Reply:
While I agree that the version you provided looks marginally prettier, it requires creating empty values on access. Since that template is to be used first with `any_ref`, I found implementing this oeprator unnecessary.

Reply:
Makes sense, ignore this then